### PR TITLE
Tidy up app loading, pt2

### PR DESCRIPTION
**Pull request by shauns**
### WHY are these changes introduced?

Relates to https://github.com/Shopify/develop-app-management/issues/1791

Continuing with cleaning up app assembly steps. This PR mainly focuses on the `link` service, which is re-used in various contexts.

### WHAT is this pull request doing?

- Rename transform/transformReverse/simplify to something clearer to low-context contributors
- Large refactor of `link` (and a smaller one of `use`)
- Updated the extension specs built into the test harnesses
- Dropped the `EmptyApp` concept -- now any `App` is a real app!
- Fixed a bug where app's tests would loop forever in watch mode -- the temp directory used for loader tests was _inside_ the source code. oops
- Ripped all API use out of the prompt functions

### How to test your changes?

Tests, TS, Linting -- there's no new functionality here

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/services/app/config/link.test.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
-import link, {LinkOptions, emptyApp} from './link.js'
-import {saveCurrentConfig} from './use.js'
+import link, {LinkOptions} from './link.js'
```
Comment (shauns) [inline packages/app/src/cli/services/app/config/link.test.ts:1]: This is a big diff section -- its just asserting existing the result of `link` for lots of these tests, and moving the content assertion to the end (as its less actionable than the other assertions)

File: packages/app/src/cli/services/app/config/link.ts [hunk 2]
```diff
@@ -1,31 +1,29 @@
-import {saveCurrentConfig} from './use.js'
+import {setCurrentConfigPreference} from './use.js'
```
Comment (shauns) [inline packages/app/src/cli/services/app/config/link.ts:1]: This is changed a lot. It's doing the same thing but in a hopefully saner order. I'd read the new version and ignore the diff

File: packages/app/src/cli/models/app/app.ts [hunk 3]
```diff
@@ -82,6 +87,9 @@ export type CurrentAppConfiguration = BasicAppConfigurationWithoutModules & Spec
  */
 export type LegacyAppConfiguration = zod.infer<typeof LegacyAppSchema> & {path: string}
 
+/** Validation schema that produces a provided app configuration type */
+type SchemaForConfig<TConfig extends {path: string}> = zod.ZodType<Omit<TConfig, 'path'>>
```
[the remark this answers is not in the document]
Comment (shauns) [reply packages/app/src/cli/models/app/app.ts:91]: I've looked at this and I think worth pulling out of the PR. It'll be really noisy, judging by the TS errors I get. I'm also not sure if things are improved by making the change. I agree with the intention though

**Review by isaacroldan — APPROVED** [review]

File: packages/app/src/cli/models/app/app.ts [hunk 4]
```diff
@@ -376,6 +376,14 @@ export class App<
     return Boolean(frontendConfig || backendConfig)
   }
 
+  creationDefaultOptions(): AppCreationDefaultOptions {
+    return {
+      isLaunchable: this.appIsLaunchable(),
+      scopesArray: getAppScopesArray(this.configuration),
```
[the remark this answers is not in the document]
Comment (shauns) [reply packages/app/src/cli/models/app/app.ts:382]: Good idea 👍

**Pull request merged**